protocol/x/subaccounts/keeper/isolated_subaccount.go

```diff
@@ -0,0 +1,145 @@
+//   - a subaccount with a position in an isolated perpetual cannot have updates for other
+//     perpetuals
+//   - a subaccount with a position in a non-isolated perpetual cannot have updates for isolated
+//     perpetuals
+//   - a subaccount with no positions cannot be updated to have positions in multiple isolated
+//     perpetuals or a combination of isolated and non-isolated perpetuals
+func isValidIsolatedPerpetualUpdates(
+	settledUpdate settledUpdate,
+	perpIdToMarketType map[uint32]perptypes.PerpetualMarketType,
+) (types.UpdateResult, error) {
+	// If there are no perpetual updates, then this update does not violate constraints for isolated
+	// markets.
+	if len(settledUpdate.PerpetualUpdates) == 0 {
+		return types.Success, nil
+	}
+
+	// Check if the updates contain an update to an isolated perpetual.
+	hasIsolatedUpdate := false
+	isolatedUpdatePerpetualId := uint32(0)
+	for _, perpetualUpdate := range settledUpdate.PerpetualUpdates {
+		marketType, exists := perpIdToMarketType[perpetualUpdate.PerpetualId]
+		if !exists {
+			return types.UpdateCausedError, errorsmod.Wrap(
+				perptypes.ErrPerpetualDoesNotExist, lib.UintToString(perpetualUpdate.PerpetualId),
+			)
+		}
+
+		if marketType == perptypes.PerpetualMarketType_PERPETUAL_MARKET_TYPE_ISOLATED {
+			hasIsolatedUpdate = true
+			isolatedUpdatePerpetualId = perpetualUpdate.PerpetualId
+			break
+		}
+	}
+
+	// Check if the subaccount has a position in an isolated perpetual.
+	// Assumes the subaccount itself does not violate the isolated perpetual constraints.
+	isIsolatedSubaccount := false
+	isolatedPositionPerpetualId := uint32(0)
+	hasPerpetualPositions := len(settledUpdate.SettledSubaccount.PerpetualPositions) > 0
+	for _, perpetualPosition := range settledUpdate.SettledSubaccount.PerpetualPositions {
+		marketType, exists := perpIdToMarketType[perpetualPosition.PerpetualId]
+		if !exists {
+			return types.UpdateCausedError, errorsmod.Wrap(
+				perptypes.ErrPerpetualDoesNotExist, lib.UintToString(perpetualPosition.PerpetualId),
+			)
+		}
+
+		if marketType == perptypes.PerpetualMarketType_PERPETUAL_MARKET_TYPE_ISOLATED {
+			isIsolatedSubaccount = true
+			isolatedPositionPerpetualId = perpetualPosition.PerpetualId
+			break
+		}
+	}
+
+	// A subaccount with a perpetual position in an isolated perpetual cannot have updates to other
+	// non-isolated perpetuals.
+	if isIsolatedSubaccount && !hasIsolatedUpdate {
+		return types.ViolatesIsolatedSubaccountConstraints, nil
+	}
+
+	// A subaccount with perpetual positions in non-isolated perpetuals cannot have an update
+	// to an isolated perpetual.
+	if !isIsolatedSubaccount && hasPerpetualPositions && hasIsolatedUpdate {
+		return types.ViolatesIsolatedSubaccountConstraints, nil
+	}
+
+	// There cannot be more than a single perpetual update if an update to an isolated perpetual
+	// exists in the slice of perpetual updates.
+	if hasIsolatedUpdate && len(settledUpdate.PerpetualUpdates) > 1 {
```

Comment:
Is it possible we have multiple perpetual update entries with the same isolated perpetual?